Api/Core/InteractiveAction.cs

```diff
@@ -0,0 +1,66 @@
+﻿using System;
+using System.Composition;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+
+namespace NosAyudamos
+{
+
+    [Export]
+    class InteractiveAction
+    {
+        readonly IRepositoryFactory repositoryFactory;
+        readonly IEnvironment environment;
+
+        public InteractiveAction(IRepositoryFactory repositoryFactory, IEnvironment environment) => (this.repositoryFactory, this.environment) = (repositoryFactory, environment);
+
+        [SuppressMessage("Design", "CA1031:Do not catch general exception types", Justification = "Catching general exeception because if retry pattern")]
+        public async Task<TResult> ExecuteAsync<TResult>(
+            Func<Task<TResult>> action,
+            Func<Task> onRetry,
+            Func<Task> onCancel,
+            [CallerMemberName] string actionName = "") where TResult : new()
+        {
+            var actionId = typeof(InteractiveAction).Name;
+            var actionRetryRepository = repositoryFactory.Create<ActionRetryEntity>();
+            var actionRetry = await actionRetryRepository.GetAsync(actionId, actionName);
+
+            if (actionRetry != null)
+            {
+                await actionRetryRepository.DeleteAsync(actionRetry);
+            }
+            else
+            {
+                actionRetry = new ActionRetryEntity(actionId, actionName);
+            }
+
+            try
+            {
+                if (actionRetry.RetryCount < environment.GetVariable<int>("ResilientNumberOfRetries", 3))
+                {
+                    TResult result = await action();
+
+                    if (result == null)
+                    {
+                        throw new InvalidOperationException();
+                    }
+
+                    return result;
+                }
+                else
+                {
+                    await onCancel();
+                }
+            }
+            catch (Exception)
+            {
+                actionRetry.RetryCount += 1;
+                await actionRetryRepository.PutAsync(actionRetry);
+                await onRetry();
+            }
+
+            return await Task.FromResult(default(TResult)!);
```

Comment:
We dont want to throw, scenario here is a person sends an invalid id and we want to recognize it, if we cant we send back a mesage to the user ("please send it again")
Same thing 3 times
At thr 4th time we sent a new message "Sorry, someone will contact you"

Perhaps a better name for this class is "DurableActionRetry" (Durable in time)

Reply:
I like that name. My point is that the `T` on this thing should be `T?` then in the signature, since it *can* return null (instead of throwing, as you mentioned.)

Reply:
ok, will change to T?